lib/vanagon/component.rb

```diff
@@ -290,6 +290,10 @@ def get_build_dir
       end
     end
 
+    def get_dependency_hash
+      { name => { 'version' => version, 'ref' => options[:ref] }.delete_if { |k, v| !v } }
```

Comment:
Nate, that's a lot of words to say "I disagree with you and I want you to stop pushing it."

> I wouldn't necessarily assume that this method will be only used for generating JSON. We almost had it generate YAML, and it may be used by other methods that may do something different with the hash.

```ruby
>> puts YAML.dump({ 'rando' => { 'version' => '1.2.3', 'ref' => nil }})
---
rando:
  version: 1.2.3
  ref:
=> nil
>> YAML.load(YAML.dump({ 'rando' => { 'version' => '1.2.3', 'ref' => nil }}))
=> {
  "rando" => {
    "version" => "1.2.3",
        "ref" => nil
  }
}
```

They're serialization formats; by definition generators have to have some mechanism to deal with dumping values like `nil` into values that the parser than understand. Just let the values be `nil`.

> Using ! to test for nil isn't exactly obvious either, since that's usually something used for boolean values. Why not use the .nil? method in this case?

I didn't suggest just calling `#nil?` because of the way everything in Ruby is truthy except `nil` and `false`. But like I said, I wouldn't even bother trying to keep empty keys out. Filter them however you like if you have your heart set on filtering.

> I'd say we've all probably learned something from these discussions, and we're all the better for it, but time is of the essence!
>
> We're nearly done with this change, so let's stay focused and finish whatever else is necessary to get this across the line.

If you've got time to argue about it, you've got time to do it right. Things that happen in Vanagon seldom get revisited once they're in. Vanagon is consumed by entities outside of Puppet. A feature may start out with a Puppet-specific need, but that feature will still be available for folks outside of Puppet to use. So that means that there's a real cost for getting even private internal API stuff like this not-quite-right. I'm asking Britt (and by extension you) to consider that cost.